R/utils.R

```diff
@@ -1064,3 +1064,322 @@ normalize_text <- function(x, lemma = FALSE) {
+    time_orig - 24,
+    time_orig
+  )
+  return(final_time)
+}
+
+
+#' @title
+#' Chopped log normal distribution for time variables
+#'
+#' @description
+#' The function samples from a log normal using `rlnorm` to obtain time of day data in hours. Values greater than 24 are subtracted by 24 (moved to the next day) so that a real time variable is observed.
+#'
+#' @param nrow (`integer`) The number of data points to sample
+#'
+#' @param meanlog (`numeric`) The log mean for the distribution
+#'
+#' @param sdlog (`numeric`) The log standard deviation
+#'
+#' @param min (`numeric`) Optional, a minimum value to left truncate the value to; the default value is set to 0.
+#'
+#' @param max (`numeric`) Optional, a maximum value to right truncate the value to; the default value is set to 48.
+#'
+#' @param seed (`integer`) Optional, an integer for setting the seed for reproducible results.
+#'
+#'
+#' @return A numeric vector following the specified distribution.
+#'
+#' @export
+#'
+sample_time_shifted_lnorm <- function(nrow, meanlog, sdlog, min = 0, max = 48, seed = NULL) {
+  if (!is.null(seed)) {
+    set.seed(seed)
+  }
+  sample_dist <- function(nrow, meanlog, sdlog) {
+    # sampling of skewed normal distribution
+    time_orig <- rlnorm(
+      n = nrow,
+      meanlog = meanlog,
+      sdlog = sdlog
+    )
+    # times greater than 24 hours are after 12am
+    # subtract 25 to turn 12am into 00:00
+    final_time <- ifelse(time_orig >= 24,
+      time_orig - 24,
+      time_orig
+    )
+    return(final_time)
+  }
+  res <- sample_dist(nrow, meanlog, sdlog)
+  while (sum(res < min) + sum(res > max) > 0) {
+    bad_sum <- sum(res < min) + sum(res > max)
+    res[c(res < min | res > max)] <- sample_dist(bad_sum, meanlog, sdlog)
+  }
+  return(res)
+}
+
+
+#' @title
+#' Generate a data table with basic inpatient stay information. At the minimum, it will include an encounter and hospital ID, along with other information if `cohort` is included in the input.
+#'
+#' @description
+#' This function creates a data table based on the input, either based on a number of unique IDs or an existing inpatient data table. It can be used to create long format data tables based on inpatient data or to create a new data based on user specifications.
+#'
+#' @param nid (`integer`)\cr Optional, number of unique encounter IDs to simulate
+#'
+#' @param n_hospitals (`integer`)\cr Optional, number of hospitals to simulate and assign to encounter IDs
+#'
+#' @param cohort (`data.table`)\cr Optional, resembling the GEMINI "admdad" table to build the returned data table from
+#'
+#' @param include_prop (`numeric`)\cr A number between 0 and 1, for the proportion of unique rows in `cohort` to include in the final data table
+#'
+#' @param avg_repeats (`numeric`)\cr The average number of repeats per row in the final data table
+#'
+#' @param by_los (`logical`)\cr Optional, whether to assign more repeats to longer hospital stays or not
```

Comment:
Since `admission_date_time` and `discharge_date_time` are used by the function, we'll need to specify that these 2 variables are required in the input `cohort` when `by_los=TRUE`
```suggestion
#' @param by_los (`logical`)\cr Optional, whether to assign more repeats to longer hospital stays or not. Default to FALSE. When TRUE, two additional columns are required in the input `cohort` dataset - `admission_date_time` and `discharge_date_time` for calculating length of stay.
```